### PR TITLE
Interference iterators

### PR DESCRIPTION
**Pull request by attackgoat**
Hi there,

I wanted to query a `CollisionWorld` instance for interference within my camera frustum but I noticed the current setup is based on per-frame allocation, with notes that it should be an iterator instead.

These changes are:
- Remove unused/duplicate `AABB` collector code
- Add `AABB` iterator code in three places: `DBVT`, `DBVTBroadPhase`, `CollisionWorld`

Design choices:
- I chose to keep instances of the `BoundingVolume` inside the iterators, as opposed to references. In my case the bv is ephemeral and so it didn't make sense to keep it around in the calling code.
- I put a constraint on the data parameter to be `Copy`, which may be too specific.
- `DBVTBroadPhase` provides access to a boxed iterator through the `BroadPhase` trait or concrete access via the struct. I used the same function name, not sure if that's good style.

Hope this helps, I would love to contribute the remaining Ray/Point implementations and work these into the query modules as well.

Thanks!

- John

### CODE REVIEW

File: src/pipeline/broad_phase/broad_phase.rs [hunk 1]
```diff
@@ -52,8 +52,8 @@ pub trait BroadPhase<N: Real, BV, T>: Any + Sync + Send {
      * FIXME: the following are not flexible enough.
      */
     // XXX: return iterators when associated types work.
-    /// Collects every object which might intersect a given bounding volume.
-    fn interferences_with_bounding_volume<'a>(&'a self, bv: &BV, out: &mut Vec<&'a T>);
+    /// Returns an interference iterator for the given bounding volume.
+    fn interferences_with_bounding_volume<'a>(&'a self, bv: BV) -> Box<'a + Iterator<Item = ProxyHandle>>;
```
Comment (Ralith) [inline src/pipeline/broad_phase/broad_phase.rs]: Could this use impl Trait instead?

**Comment by sebcrozet** [comment]
Hi!

Thank you for this! I am not sure yet if this is the best approach.
My main concerns regarding this proposal are:

1. The `DBVTBVIterator` basically re-implements the tree traversal. It would be better to find a way to improve code-reuse. Something like the visitor pattern, but for iterators. Maybe there is a way to define a DBVT method like `pub fn visit_iterator<Vis: BVTIteratorVisitor<B, BV>>(&self, visitor: &mut Vis) -> impl Iterator<...>`. The `visit_iterator` would implement the actual tree traversal, like your `DBVTBVIterator`, and use the provided visitor to direct the traversal.
2. The users of the `BroadPhase` trait will suffer from the two allocations (of the `Vec` and `Box` for the iterator). The previous implementation had the benefit to potentially avoid allocations by letting the user providing its own `Vec`.
3. Is there a way to avoid the `CollisionWorld::proxies` hashmap ? Maintaining this map and using it at each access the iterator is expensive.

Also, it is unfortunate to now have two allocation sources: the `Vec` inside of the iterator, and the `Box<'a + Iterator<Item = ProxyHandle>>`. Though I understand we can't avoid this box because simply returning `impl Iterator<Item = ProxyHandle>` would prevent using `BroadPhase` as a trait-object. So we might not be able to do anything about this unfortunately.

Regarding your design choices:

- It would be better to keep references to the bounding volume instead of instance. In the future, we might introduce bounding volumes more complex than just `AABB` and `BoundingSphere`. For example a convex hull bounding volume would not be `Copy`.
- It would be better to keep the `Clone` constraint on `B` so it matches the previous implementation results.

**Comment by attackgoat** [comment]
1. [Generators](https://doc.rust-lang.org/std/ops/trait.Generator.html) would help, but they aren't stable yet. Two ideas:
- BVTIteratorVisitor provides a callback Fn for each result with a cancel/proceed return value.
- Just gate this feature behind `#![feature(generators)]`
2. The `vec` allocation could be solved by #1, but I don't think there is a way to avoid the `Box` until `impl` [can be used in a trait](https://github.com/rust-lang/rust/issues/34511).
3. I think this can be solved by storing the slab indices in the data parameter, or at the very least updating it once so that no hashmap is needed.

**Comment by attackgoat** [comment]
It'll be a few days but I'll address the design issues and see if I can get a better change set. Thanks!!

**Comment by aleksijuvani** [comment]
I'm wondering what the current status of this pull request is. Have the concerns above been addressed? I'm looking into implementing convex casts (issue #29) and this would likely involve adding a new method to the collision world very similar to the existing interference methods. If these functions are going to change to use iterators moving forward, I would prefer to also use iterators in the convex cast method and avoid having to rewrite the function later.

**Comment by attackgoat** [comment]
I think the approach I took will only be worthwhile once generators come around.

**Pull request closed without merging**